geode-for-redis/src/main/java/org/apache/geode/redis/internal/eventing/BlockingCommandListener.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.eventing;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.geode.annotations.VisibleForTesting;
+import org.apache.geode.redis.internal.commands.Command;
+import org.apache.geode.redis.internal.commands.RedisCommandType;
+import org.apache.geode.redis.internal.commands.executor.RedisResponse;
+import org.apache.geode.redis.internal.data.RedisKey;
+import org.apache.geode.redis.internal.netty.Coder;
+import org.apache.geode.redis.internal.netty.ExecutionHandlerContext;
+
+public class BlockingCommandListener implements EventListener {
+
+  private final ExecutionHandlerContext context;
+  private final Command command;
+  private final List<RedisKey> keys;
+  private final double timeoutSeconds;
+  private final long timeSubmitted;
+  private final AtomicBoolean active = new AtomicBoolean(true);
+
+  /**
+   * Constructor to create an instance of a BlockingCommandListener in response to a blocking
+   * command. When receiving a relevant event, blocking commands simply resubmit the command
+   * into the Netty pipeline.
+   *
+   * @param context the associated ExecutionHandlerContext
+   * @param command the blocking command associated with this listener
+   * @param keys the list of keys the command is interested in
+   * @param timeoutSeconds the timeout for the command to block in seconds
+   */
+  public BlockingCommandListener(ExecutionHandlerContext context, Command command,
+      List<RedisKey> keys, double timeoutSeconds) {
+    this.context = context;
+    this.command = command;
+    this.timeoutSeconds = timeoutSeconds;
+    this.keys = Collections.unmodifiableList(keys);
+    timeSubmitted = System.nanoTime();
+  }
+
+  @Override
+  public List<RedisKey> keys() {
+    return keys;
+  }
+
+  @Override
+  public EventResponse process(RedisCommandType commandType, RedisKey key) {
+    if (!keys.contains(key)) {
+      return EventResponse.CONTINUE;
+    }
+
+    if (active.compareAndSet(true, false)) {
```

Comment:
I think it's a good idea. This also ensures that the timeout does not fully process when a listener is resubmitted. Since the scheduled timeouts are not being cancelled we will have some listeners hanging around until their timeouts fire. Do you think that's an issue? I could add a `cleanup()` call to the listener that is executed when the listener is removed.